webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/RepositoryModal.js

```diff
@@ -0,0 +1,180 @@
+  repoData,
+  handleRepoData,
+  isLoading,
+  isModalOpen,
+  handleModalToggle,
+}) => {
+  const initialRepoState = [
+    {
+      reference: 0,
+      repository: '',
+      gpgKeyUrl: '',
+    },
+  ];
+
+  useEffect(() => {
+    if (!repoData.length) {
+      handleRepoData(initialRepoState);
+    }
+  }, [repoData, initialRepoState, handleRepoData]);
+
+  const [repoReferenceCounter, setRepoReferenceCounter] = useState(0);
+
+  const updateInput = (reference, input, inputType) => {
+    handleRepoData(current =>
+      current.map(repo => {
+        if (repo.reference === reference) {
+          if (inputType === 'repository') {
+            return { ...repo, repository: input };
+          }
+
+          return { ...repo, gpgKeyUrl: input };
+        }
+        return repo;
+      })
+    );
+  };
+
+  const removeRepository = reference => {
+    handleRepoData(current =>
+      current.filter(repo => repo.reference !== reference)
+    );
+  };
+
+  const addRepositoryModalInput = () => {
+    const newRepoReferenceCounter = repoReferenceCounter + 1;
+    handleRepoData(current => [
+      ...current,
+      {
+        reference: newRepoReferenceCounter,
+        repository: '',
+        gpgKeyUrl: '',
+      },
+    ]);
+    setRepoReferenceCounter(newRepoReferenceCounter);
+  };
+
+  const clearRepositories = () => {
+    handleRepoData(initialRepoState);
+    handleModalToggle();
+  };
+
+  const renderRepo = repo => (
+    <Split hasGutter key={'repo_modal_list'.concat(repo.reference.toString())}>
+      <SplitItem span={10}>
+        <RepositoryModalInput
+          reference={repo.reference}
+          updateRepository={updateInput}
+          isLoading={isLoading}
+          initRepo={repo.repository}
+          initGpgKey={repo.gpgKeyUrl}
+        />
+      </SplitItem>
+      <SplitItem>
+        <Button
+          ouiaId={repo.reference}
+          variant="link"
+          icon=<MinusCircleIcon />
+          onClick={() => removeRepository(repo.reference)}
+        >
+          {__('Remove')}
+        </Button>
+      </SplitItem>
+    </Split>
+  );
+
+  return (
+    <Modal
+      ouiaId="host_reg_repo_modal"
+      key="host_reg_repo_modal"
+      title={__('Repository list')}
+      isOpen={isModalOpen}
+      onClose={clearRepositories}
+      actions={[
+        <Button
+          ouiaId="reg_modal_confirm"
+          key="reg_modal_confirm"
+          variant="primary"
+          onClick={handleModalToggle}
+        >
+          {__('Confirm')}
+        </Button>,
+        <Button
+          ouiaId="reg_modal_cancel"
+          key="reg_modal_cancel"
+          variant="link"
+          onClick={clearRepositories}
+        >
+          {__('Cancel')}
+        </Button>,
+        <Button
+          ouiaId="reg_modal_add_new_repo"
+          key="reg_modal_add_new_repo"
+          variant="link"
+          icon={<PlusCircleIcon />}
+          onClick={addRepositoryModalInput}
+        >
+          {__('Add repository')}
```

Comment:
Currently, I can add new fields and leave them blank and it doesn't complain or remove them. The set count is correct but the Modal shouldn't have blank rows.

Reply:
Addition: This doesn't affect the rendered `global_registration.erb` template: If there are 2 repositories and 2 blank lines only to repositories are added during host registration.

Reply:
Yes, the empty rows are visible in the UI but they are not sent to the backend.  I am unsure what to do about the empty rows in the UI. I think it might confuse the user if I remove them when they click confirm and then re-open the modal for editing.

Reply:
I think it makes sense to remove empty rows after submitting

Reply:
K, fixed it.